runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -235,31 +251,33 @@ public int appendObjectState(Stream<Map.Entry> entryStream) {
         ImmutableMap<CheckpointEntry.CheckpointDictKey, String> mdkv =
                 ImmutableMap.copyOf(this.mdkv);
 
-        final Iterable<List<Map.Entry>> partitions =
-                Iterables.partition(entryStream::iterator, batchSize);
         int entryCount = 0;
-
-        for (List<Map.Entry> partition : partitions) {
-            MultiSMREntry smrEntries = new MultiSMREntry();
-            for (Map.Entry entry : partition) {
-                smrEntries.addTo(new SMREntry("put",
-                        new Object[]{keyMutator.apply(entry.getKey()),
-                                valueMutator.apply(entry.getValue())},
-                        serializer));
-                entryCount++;
+        int numBytesPerCheckpointEntry = 0;
+        ByteBuf b = Unpooled.buffer();
+        MultiSMREntry smrEntries = new MultiSMREntry();
+
+        for (Map.Entry entry : (Iterable<Map.Entry>) entryStream::iterator) {
+            SMREntry each = new SMREntry("put",
+                    new Object[]{keyMutator.apply(entry.getKey()),
+                            valueMutator.apply(entry.getValue())},
+                    serializer);
+
+            each.serialize(b);
+            numBytesPerCheckpointEntry += b.writerIndex();
+
+            if(numBytesPerCheckpointEntry > rt.getParameters().getMaxWriteSize()) {
+                convertToCheckpointEntry(smrEntries);
+                numBytesPerCheckpointEntry = b.writerIndex();
```

Comment:
Is this `numBytesPerCheckpointEntry = 0;`?

Reply:
You're right. Got it.

Reply:
b.writerIndex() will not have the correct value as the last entry 'each' has already been serialized inside it, right?

Reply:
clarified offline